src/models/document/document.ts

```diff
@@ -108,6 +108,22 @@ export const DocumentModel = types
     }
   }))
   .views(self => ({
+    matchProperties(properties: string[]) {
+      return properties.every(p => {
+        const match = /(!)?(.*)/.exec(p);
+        const property = match && match[2];
+        const wantsProperty = !(match && match[1]); // not negated => has property
+        // treat "starred" as a virtual property
+        if (property === "starred") {
+          return self.isStarred === wantsProperty;
+        }
+        if (property) {
+          return !!self.getProperty(property) === wantsProperty;
+        }
+        // ignore empty strings, etc.
+        return true;
+      });
+    },
```

Comment:
Implementation moved from the `DocumentsSection` component essentially unchanged.